parsers/openapi.c

```diff
@@ -0,0 +1,363 @@
+	KEY_UNKNOWN,
+	KEY_PARAMETERS,
+};
+
+static const enum openapiKeys schemas3Keys[] = {
+	KEY_UNKNOWN,
+	KEY_SCHEMAS,
+	KEY_COMPONENTS,
+};
+
+static const enum openapiKeys definitions2Keys[] = {
+	KEY_UNKNOWN,
+	KEY_DEFINITIONS,
+};
+
+const struct tagSource tagSources[] = {
+	{
+		KIND_PATH,
+		pathKeys,
+		ARRAY_SIZE (pathKeys),
+	},
+	{
+		KIND_RESPONSE,
+		responses3Keys,
+		ARRAY_SIZE (responses3Keys),
+	},
+	{
+		KIND_RESPONSE,
+		responses2Keys,
+		ARRAY_SIZE (responses2Keys),
+	},
+	{
+		KIND_PARAMETER,
+		parameters3Keys,
+		ARRAY_SIZE (parameters3Keys),
+	},
+	{
+		KIND_PARAMETER,
+		parameters2Keys,
+		ARRAY_SIZE (parameters2Keys),
+	},
+	{
+		KIND_SCHEMA,
+		schemas3Keys,
+		ARRAY_SIZE (schemas3Keys),
+	},
+	{
+		KIND_SCHEMA,
+		definitions2Keys,
+		ARRAY_SIZE (definitions2Keys),
+	}
+};
+
+static void handleKey(struct sOpenApiSubparser *openapi,
+											 yaml_token_t *token)
+{
+	int i;
+	for (i = 0; i < ARRAY_SIZE(tagSources); i++) {
+		const struct tagSource* ts = &tagSources[i];
+
+		if (stateStackMatch(
+					openapi->type_stack,
+					ts->keys,
+					ts->countKeys
+					)) {
+
+			tagEntryInfo tag;
+			initTagEntry (&tag, (char *)token->data.scalar.value, ts->kind);
+			attachYamlPosition (&tag, token, false);
+
+			makeTagEntry (&tag);
+		}
+	}
+}
+
+static void	openapiPlayStateMachine (struct sOpenApiSubparser *openapi,
+											 yaml_token_t *token)
+{
+	// printStack(openapi->type_stack);
+
+	switch (token->type)
+	{
+	case YAML_KEY_TOKEN:
+		openapi->play_detection_state = DSTAT_LAST_KEY;
+		break;
+	case YAML_SCALAR_TOKEN:
+		switch (openapi->play_detection_state) {
+			case DSTAT_LAST_KEY:
+				openapi->type_stack->key = parseKey(token);
+				//printf("  key: %s\n", (char*)token->data.scalar.value);
+				handleKey (openapi, token);
+				break;
+			case DSTAT_LAST_VALUE:
+				//printf("  value: %s\n", (char*)token->data.scalar.value);
+				break;
+			default:
+				break;
+		}
+
+		openapi->play_detection_state = DSTAT_INITIAL;
+
+		break;
+	case YAML_VALUE_TOKEN:
+		openapi->play_detection_state = DSTAT_LAST_VALUE;
+		break;
+
+	default:
+		openapi->play_detection_state = DSTAT_INITIAL;
+		break;
+	}
+}
+
+static void newTokenCallback (yamlSubparser *s, yaml_token_t *token)
+{
+	if (token->type == YAML_BLOCK_SEQUENCE_START_TOKEN
+		|| token->type == YAML_BLOCK_MAPPING_START_TOKEN)
+		pushBlockType ((struct sOpenApiSubparser *)s, token->type);
+
+	openapiPlayStateMachine ((struct sOpenApiSubparser *)s, token);
+
+	if (token->type == YAML_BLOCK_END_TOKEN)
+		popBlockType ((struct sOpenApiSubparser *)s, token);
+	else if (token->type == YAML_STREAM_END_TOKEN)
+		popAllBlockType ((struct sOpenApiSubparser *)s, token);
+}
+
+static void inputStart(subparser *s)
+{
+	((struct sOpenApiSubparser*)s)->play_detection_state = DSTAT_INITIAL;
+	((struct sOpenApiSubparser*)s)->type_stack = NULL;
+}
+
+static void inputEnd(subparser *s)
+{
+	Assert (((struct sOpenApiSubparser*)s)->type_stack == NULL);
```

Comment:
Even if a broken input is given to the parser, we can expect the type_stack becomes empty at the end of parsing?

If we cannot expect it, we should clear the stack here instead of doing "Assert".

Reply:
ok, done